sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala

```diff
@@ -62,20 +61,113 @@ case class CustomShuffleReaderExec private(
     }
   }
 
-  override def stringArgs: Iterator[Any] = Iterator(description)
+  override def stringArgs: Iterator[Any] = {
+    val desc = if (isLocalReader) {
+      "local"
+    } else {
+      if (hasCoalescedPartition && hasSkewedPartition) {
+        "coalesced and skewed"
+      } else if (hasCoalescedPartition) {
+        "coalesced"
+      } else if (hasSkewedPartition) {
+        "skewed"
+      } else {
+        ""
+      }
+    }
+    Iterator(desc)
+  }
 
-  private var cachedShuffleRDD: RDD[InternalRow] = null
+  def hasCoalescedPartition: Boolean = {
```

Comment:
Is this visible for testing or do you have some use cases in your mind?
`hasCoalescedPartition`, `hasSkewedPartition`, `isLocalReader`?

Reply:
It's also used in `stringArgs`.

coalesced partition and skewed partition can co-exist, but local reader partition is exclusive.